spinoso-string/src/enc/ascii/mod.rs

```diff
@@ -280,6 +281,18 @@ impl AsciiString {
     pub fn extend_from_slice(&mut self, other: &[u8]) {
         self.inner.extend_from_slice(other);
     }
+
+    #[inline]
+    pub fn codepoints(&self) -> Codepoints {
+        let iter = self
+            .inner
+            .as_slice()
+            .bytes()
+            .map(char::from)
```

Comment:
I'm not sure it is correct to call `char::from`. What is the behavior when the encoding is not valid?

```console
$ irb
[3.2.2] > s = "\0\xFF"
=> "\u0000\xFF"
[3.2.2] > s.codepoints
(irb):2:in `codepoints': invalid byte sequence in UTF-8 (ArgumentError)
        from (irb):2:in `<main>'
        from /usr/local/var/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `<main>'
[3.2.2] > s.dup.b.codepoints
=> [0, 255]
[3.2.2] > s.dup.force_encoding('ascii').codepoints
=> [0, 255]
```

Reply:
I reverted to use `u32::from` instead of `char::from`.

Also tested this for ASCII in e6be81d and asserted invalid bytes are just interpreted as bytes (e.g. `\xFF` -> `255`) instead of raising error:

https://github.com/artichoke/artichoke/blob/e6be81d1dbd798b8215f7d4031f83611b6ff9b8b/spinoso-string/src/enc/ascii/codepoints.rs#L52-L56

However, from `airb` an invalid UTF-8 code point error is still raised even with `force_encoding('ascii')`, as of e6be81d1dbd798b8215f7d4031f83611b6ff9b8b:
```console
$ cargo run --bin airb
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target/debug/airb`
artichoke 0.1.0-pre.0 (2023-10-20 revision 7021) [aarch64-apple-darwin]
[rustc 1.70.0 (90c541806 2023-05-31) on aarch64-apple-darwin]
>>> "hello".codepoints
=> [104, 101, 108, 108, 111]
>>> "hello💎".codepoints
=> [104, 101, 108, 108, 111, 128142]
>>>
>>> s = "\0\xFF"
=> "\x00\xFF"
>>> s.codepoints
Traceback (most recent call last):
	2: from (airb):1
	1: from (airb):1:in codepoints
ArgumentError (invalid byte sequence in UTF-8)
>>> s.dup.b.codepoints
=> [0, 255]
>>> s.dup.force_encoding('ascii').codepoints
Traceback (most recent call last):
	2: from (airb):1
	1: from (airb):1:in codepoints
ArgumentError (invalid byte sequence in UTF-8)
>>>
```

Since `map(u32::from)` is introduced before this PR, perhaps that's related to another issue #1726?

Reply:
ahh yea, `String#force_encoding` is a stub right now, not properly implemented:

https://github.com/artichoke/artichoke/blob/f32bd63872b8b5b247c3bcb6516801767290057f/artichoke-backend/src/extn/core/string/string.rb#L519-L526

I think getting that fixed is part of @b-n's work on implementing `Encoding`